wsmux/circ_buffer.go

```diff
@@ -0,0 +1,77 @@
+package wsmux
+
+type buffer struct {
+	buf   []byte
+	s     int  // start: points to first byte containing data
+	e     int  // end: points to last byte containing data + 1
+	c     int  // capacity
+	empty bool // required : s == e can be because buffer is empty or buffer is full
```

Comment:
So instead of storing `start` and `end` how about storing `start` and `size`.

Then you won't need this boolean... it might reduce the number special cases...